src/gen_shader_bin.c

```diff
@@ -0,0 +1,300 @@
+        fprintf(out, "    0x%08X,\n", asm_alu_instructions[i].part4);
+        fprintf(out, "    0x%08X,\n", asm_alu_instructions[i].part7);
+        fprintf(out, "    0x%08X,\n", asm_alu_instructions[i].part6);
+    }
+
+    fprintf(out, "    HOST1X_OPCODE_NONINCR(0x806, %d),\n",
+            asm_fs_instructions_nb);
+    for (i = 0; i < asm_fs_instructions_nb; i++)
+        fprintf(out, "    0x%08X,\n", asm_alu_instructions[i].complement);
+
+    fprintf(out, "    HOST1X_OPCODE_NONINCR(0x901, %d),\n",
+            asm_fs_instructions_nb);
+    for (i = 0; i < asm_fs_instructions_nb; i++)
+        fprintf(out, "    0x%08X,\n", asm_dw_instructions[i].data);
+
+    fprintf(out, "};\n\n");
+
+    fprintf(out, "static uint32_t vs_%s_words[] = {\n", fp_name);
+
+    fprintf(out, "    HOST1X_OPCODE_NONINCR(0x206, %d),\n",
+           asm_vs_instructions_nb * 4);
+
+    for (i = 0; i < asm_vs_instructions_nb; i++) {
+        fprintf(out, "    0x%08X,\n", asm_vs_instructions[i].part3);
+        fprintf(out, "    0x%08X,\n", asm_vs_instructions[i].part2);
+        fprintf(out, "    0x%08X,\n", asm_vs_instructions[i].part1);
+        fprintf(out, "    0x%08X,\n", asm_vs_instructions[i].part0);
+    }
+
+    fprintf(out, "};\n\n");
+
+    for (i = 0; i < 16; i++) {
+        if (asm_vs_attributes[i].used)
+            in_mask |= 1 << i;
+
+        if (asm_vs_exports[i].used)
+            out_mask |= 1 << i;
+    }
+
+    fprintf(out, "static uint32_t lnk_%s_words[] = {\n", fp_name);
+
+    fprintf(out, "    HOST1X_OPCODE_INCR(0x300, %d),\n",
+           asm_linker_instructions_nb * 2);
+
+    for (i = 0; i < asm_linker_instructions_nb; i++) {
+        fprintf(out, "    0x%08X,\n", asm_linker_instructions[i].first);
+        fprintf(out, "    0x%08X,\n", asm_linker_instructions[i].latter);
+    }
+
+    fprintf(out, "};\n\n");
+
+    fprintf(out, "static struct shader_program prog_%s = {\n", fp_name);
+    fprintf(out, "    .vs_prog_words = vs_%s_words,\n", fp_name);
+    fprintf(out, "    .vs_prog_words_nb = TEGRA_ARRAY_SIZE(vs_%s_words),\n", fp_name);
+    fprintf(out, "    .vs_attrs_in_mask = %u,\n", in_mask);
+    fprintf(out, "    .vs_attrs_out_mask = %u,\n", out_mask);
+    fprintf(out, "\n");
+    fprintf(out, "    .fs_prog_words = fs_%s_words,\n", fp_name);
+    fprintf(out, "    .fs_prog_words_nb = TEGRA_ARRAY_SIZE(fs_%s_words),\n", fp_name);
+    fprintf(out, "    .fs_alu_buf_size = %u,\n", asm_alu_buffer_size);
+    fprintf(out, "    .fs_pseq_to_dw = %u,\n", asm_pseq_to_dw_exec_nb);
+    fprintf(out, "    .fs_pseq_inst_nb = %u,\n", asm_fs_instructions_nb);
+    fprintf(out, "\n");
+    fprintf(out, "    .linker_words = lnk_%s_words,\n", fp_name);
+    fprintf(out, "    .linker_words_nb = TEGRA_ARRAY_SIZE(lnk_%s_words),\n", fp_name);
+    fprintf(out, "    .linker_inst_nb = %u,\n", asm_linker_instructions_nb);
+    fprintf(out, "    .used_tram_rows_nb = %u,\n", asm_linker_used_tram_rows_nb);
+    fprintf(out, "};\n");
+
+    return 0;
+}
```

Comment:
I have a feeling that introducing tools like this should come as a separate commit. That would also make it a bit easier to review.

Reply:
Ok